src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,219 @@
+    val jmsConnector = new JmsConnector(config)
+
+    val result = for {
+      q <- localQueue
+      res <- jmsConnector.getJmsIo(QueueName(config.requestQueue)).use { case (jmsProducer, jmsConsumer) =>
+               process(q, jmsConsumer, jmsProducer)
+             }
+    } yield res
+    result.as(ExitCode.Success)
+  }
+
+  private def process(q: Queue[IO, LocalMessage], consumer: JmsAcknowledgerConsumer[IO], producer: JmsProducer[IO]) =
+    for {
+      // START
+      res <- IO.race(
+               transferMessageToLocalQueue(q)(consumer),
+
+               // single dispatcher
+               //        dispatcher(q).foreverM
+
+               // multiple dispatchers
+               List.range(start = 0, end = config.maxDispatchers).parTraverse_ { i =>
+                 IO.println(s"Starting consumer #${i + 1}") >>
+                   new Dispatcher(new LocalProducerImpl(producer, QueueName(config.replyQueue))).run(i)(q).foreverM
+               }
+             )
+      // END
+
+    } yield res
+
+  private def doStop(): Unit =
+    logger.info("Connection closed")
+
+  private def acknowledgeMessage(): IO[AckAction[IO]] =
+    //    IO.pure(AckAction.ack)
+    IO {
+      println("Acknowledged message")
+      AckAction.ack
+    }
+
+  private def createLocalMessage(persistentMessageId: PersistentMessageId, jmsMessage: JmsMessage): IO[LocalMessage] =
+    for {
+      sid       <- IO.fromOption(jmsMessage.getStringProperty("sid"))(throw new Exception("Missing service ID"))
+      serviceId <- IO.fromOption(ServiceIdentifier.withNameOption(sid))(throw new Exception("SID not recognised"))
+      messageId <- IO.fromOption(jmsMessage.getJMSMessageId)(throw new Exception("Missing message ID"))
+      text      <- jmsMessage.asTextF[IO]
+    } yield LocalMessage(persistentMessageId, serviceId, text, messageId)
+
+  // receives a message, persists it and then adds it to the queue
+  private def transferMessageToLocalQueue(
+    q: Queue[IO, LocalMessage]
+  )(jmsAcknowledgerConsumer: JmsAcknowledgerConsumer[IO]): IO[Unit] =
+    jmsAcknowledgerConsumer.handle { (jmsMessage, mf) =>
+      for {
+        persistentMessageId <- localMessageStore.persistMessage(jmsMessage)
+        res                 <- acknowledgeMessage()
+        localMessage        <- createLocalMessage(persistentMessageId, jmsMessage)
+        _                   <- q.offer(localMessage)
+        _ <- IO.delay {
+               println(s"Queued: $persistentMessageId")
+             }
+      } yield res
+    }
+
+}
+object ApiService {
+
+  implicit val loggerFactory: LoggerFactory[IO] = Slf4jFactory[IO]
+  implicit val logger: SelfAwareStructuredLogger[IO] = LoggerFactory[IO].getLogger
+  def apply(args: ServiceArgs): ApiService = {
+    val serviceConfig = loadConfig(args)
+    new ApiService(serviceConfig)
+  }
+
+  private def loadConfig(serviceArgs: ServiceArgs): ServiceConfig = {
```

Comment:
This is a little confusing - what's happening in this method, and when would we want to load config that isn't in `application.conf`?

Reply:
This has been removed now.